internal/messagejournal/messagejournal.go

```diff
@@ -0,0 +1,216 @@
+// which gets cleared at the start of every session and the persistent table
+// which maintains message entries across multiple sessions.
+func (j *MessageJournal) AddEntry(entry yggdrasil.WorkerMessage) (*yggdrasil.WorkerMessage, error) {
+	// If the message journal is nil, it means the message journal
+	// is not enabled in the config.toml file or provided as a user argument.
+	if j == nil {
+		return nil, fmt.Errorf("message journal not enabled")
+	}
+
+	runtimeAction, err := j.database.Prepare("INSERT INTO runtime(message_id, sent, worker_name, response_to, worker_event, worker_message) values (?,?,?,?,?,?)")
+	if err != nil {
+		return nil, err
+	}
+	persistentAction, err := j.database.Prepare("INSERT INTO persistent(message_id, sent, worker_name, response_to, worker_event, worker_message) values (?,?,?,?,?,?)")
+	if err != nil {
+		return nil, err
+	}
+
+	runtimeResult, err := runtimeAction.Exec(entry.MessageID, entry.Sent, entry.WorkerName, entry.ResponseTo, entry.WorkerEvent.EventName, entry.WorkerEvent.EventMessage)
+	if err != nil {
+		return nil, err
+	}
+	persistentResult, err := persistentAction.Exec(entry.MessageID, entry.Sent, entry.WorkerName, entry.ResponseTo, entry.WorkerEvent.EventName, entry.WorkerEvent.EventMessage)
+	if err != nil {
+		return nil, err
+	}
+
+	runtimeEntryID, err := runtimeResult.LastInsertId()
+	if err != nil {
+		return nil, fmt.Errorf("%v-%v", runtimeEntryID, err)
+	}
+	presistentEntryID, err := persistentResult.LastInsertId()
+	if err != nil {
+		return nil, fmt.Errorf("%v-%v", presistentEntryID, err)
+	}
+
+	return &entry, nil
+}
+
+func (j *MessageJournal) GetEntries(filter MessageJournalFilter) ([]map[string]string, error) {
+	// If the message journal is nil, it means the message journal
+	// is not enabled in the config.toml file or provided as a user argument.
+	if j == nil {
+		return nil, nil
+	}
+
+	entries := []map[string]string{}
+
+	// Build SQL query to retrieve journal entries.
+	// FIXME: It works but I hate it... is there a better
+	// way to do this without an external library?
+	args := []interface{}{}
+	queryString := "SELECT * FROM "
+	if filter.Persistent {
+		queryString += "persistent "
+	} else {
+		queryString += "runtime "
+	}
+	if filter.MessageID != "" || filter.Worker != "" || filter.From != "" || filter.To != "" {
+		queryString += "WHERE "
+	}
+	if filter.MessageID != "" {
+		queryString += "message_id = $1 "
+		args = append(args, filter.MessageID)
+		if filter.Worker != "" {
+			queryString += "AND "
+		}
+	}
+	if filter.Worker != "" {
+		queryString += "worker_name = $2 "
+		args = append(args, filter.Worker)
+		if filter.From != "" {
+			queryString += "AND "
+		}
+	}
+	if filter.From != "" {
+		queryString += "sent >= $3 "
+		args = append(args, filter.From)
+		if filter.To != "" {
+			queryString += "AND "
+		}
+	}
+	if filter.To != "" {
+		queryString += "sent <= $4 "
+		args = append(args, filter.To)
```

Comment:
Is there a way to create dynamic sql query strings in Go without bringing in an ORM or other external library? It looks messy constructing it with a series of conditional branches...